lib/charms/loki_k8s/v1/loki_push_api.py

```diff
@@ -349,6 +352,72 @@ def _promtail_error(self, event):
     )
 ```
 
+## LogForwarder Library Usage
+
+Let's say that we have a workload charm that produces logs to the standard output (stdout),
+and we need to send those logs to a workload implementing the `loki_push_api` interface,
+such as `Loki` or `Grafana Agent`.
+
+Use the `LogForwarder` class by instantiating it in the `__init__` method of the charmed
+operator. There are two ways to provide Loki endpoint(s) to the forwarder. You can let the
+object extract the endpoint(s) from relation data, or you can explicitly pass the Loki
+endpoint(s) to forward your logs to.
+
+1. If you want your charm to relate to another implementing the `loki_push_api` interface,
+   you only need to instantiate the object; for example:
+
+   ```python
+   from charms.loki_k8s.v1.loki_push_api import LogForwarder
+
+   ...
+
+      def __init__(self, *args):
+          ...
+          self._log_forwarder = LogForwarder(self)
+   ```
+
+   The `LogForwarder` will listen to relation events out-of-the-box and enable or disable
+   the log forwarding accordingly. Next, modify the `metadata.yaml` file to add:
+
+   - The `log-forwarding` relation in the `requires` section:
+     ```yaml
+     requires:
+       log-forwarding:
+         interface: loki_push_api
+         optional: true
+     ```
+
+2. If you don't want to relate your charm to another implementing the `loki_push_api` interface,
+   you need to explicitly provide the endpoint(s) to forward your logs to. If your charm receives
+   the endpoint(s) from another different relation, this is the approach to follow.
+   However, you also need to manually enable and disable the log forwarding. For example,
+   let's say the charm gets the endpoint(s) from the `foo` relation:
+
+   ```python
+   from charms.loki_k8s.v1.loki_push_api import LogForwarder
+
+   ...
+
+     def __init__(self, *args):
+         ...
+         self._log_forwarder = LogForwarder(
+             self,
+             loki_endpoints=self.get_loki_endpoints,
+         )
+         self.framework.observe(self.on["foo"].relation_joined, self._log_forwarder.enable())
```

Comment:
I think this API could be improved by: passing to logforwarder the endpoint which it should listen to, and a Callable[[Relation], endpoint].
Since it seems that what you're doing with the relation events is generic, and what is specific is only 'how to extract the data I need from a relation'. 

not sure I'm explaining myself well here, contact me for more.
In short:

```
  self._log_forwarder = LogForwarder(
             self,
             binding="other-endpoint", # defaults to 'logging'
             loki_endpoint_getter=self._get_loki_endpoints, # defaults to whatever we need to do to get the endpoint from logging:loki_push_api
         )

def _get_loki_endpoints(self, relation:Relation) -> str:
    return relation.data[relation.app]['very-specific-way-to-store-an-url']

```

Reply:
This is cool, done!